### PR TITLE
ns for fx: preserve observers and fake_quants through passes

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61380
* #61377
* #61360
* #61359
* #61358
* #61357
* **#61323**
* #61129

Summary:

Before this PR, all observers and fake quants were silently removed
when adding loggers with NS. This is problematic for QAT models because
we need the fake quants to run in order to properly capture intermediate
outputs.

This PR fixes the issue by preserving the observers throughout
the passes which add loggers.  In detail:
* for each quantization module or fusion, add additional patterns with that fusion and an observer/fake_quant at the end
* remove the places in the logger model creation code which removed observers
* add unit testing that QAT numerics do not change after adding loggers

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_loggers_preserve_qat_numerics
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_shadow_loggers_preserve_qat_numerics
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29600351](https://our.internmc.facebook.com/intern/diff/D29600351)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600351).

File: torch/quantization/ns/pattern_utils.py [hunk 1]
```diff
@@ -65,9 +69,23 @@ def get_reversed_fusions() -> Set[Tuple[NSFusionType, int]]:
...
+        # without caring about fusions).
         if isinstance(quant_pattern, tuple):
-            results.add((quant_pattern, default_base_op_idx))  # type: ignore[arg-type]
+            results.append((quant_pattern, default_base_op_idx))  # type: ignore[arg-type]
+
+        # For each pattern, add additional patterns with observers and
+        # fake quants at the end.
+        # TODO(future PR): if needed, implement matching for a node
+        #   having multiple output observers.
+        for cls in (ObserverBase, FakeQuantizeBase):
```
Comment (hx89) [inline torch/quantization/ns/pattern_utils.py:82]: Would there be the cases where both input and output of the op have observers or fake_quant attached, or observers and fake_quant only appear at the op output? My impression is for QAT we attach fake_quant to both input and output?

Comment (jerryzh168) [reply torch/quantization/ns/pattern_utils.py:82]: yeah I think so, is the current pattern defined as fusion_pattern + fake_quantize/observer at the output?

Comment (vkuzo) [reply torch/quantization/ns/pattern_utils.py:82]: here is a way to think about it

```
# fp32
x0 -> conv0 -> x1

# prepared
x0 -> fq_of_x0 -> conv0 -> fq_of_conv0 -> x1

# prepared, with input and output loggers
x0 -> fq_of_x0 -> input_logger_of_conv0 -> conv0 -> fq_of_conv0 -> output_logger_of_conv0 -> x1
```

in this way of thinking about it, the input of the previous op is a black box and it shouldn't matter whether it has observers or not.  We can insert input and output loggers around the current op and its observer, and it should work as needed.

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600351).

File: torch/quantization/ns/pattern_utils.py (hunk 1 above)
Comment (hx89) [reply torch/quantization/ns/pattern_utils.py:82]: Thanks for the explanation! It makes more sense now. 

One case I'm thinking is in QAT we could add fake_quant to both input and output of the op, and want to insert logger around them, for exmaple:
```
# prepared 
x0 -> fq_in_of_conv0 -> conv0 -> fq_out_of_conv0 -> x1

# prepared, with input and output loggers 
x0 ->  input_logger_of_conv0->fq_in_of_conv0 -> conv0 -> fq_out_of_conv0->output_logger_of_conv0 -> x1
```
is it possible to make this function more flexible to handle above case?

**Review by hx89 — APPROVED** [review]

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600351).

**Pull request closed without merging**